modules/OFConnectionManager2/module/src/cxn_instance.c

```diff
@@ -0,0 +1,2122 @@
+/****************************************************************
+ *
+ *        Copyright 2013-2015, Big Switch Networks, Inc.
+ *
+ * Licensed under the Eclipse Public License, Version 1.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at
+ *
+ *        http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied. See the License for the specific
+ * language governing permissions and limitations under the
+ * License.
+ *
+ ****************************************************************/
+
+
+/**
+ * @file
+ * @brief Routines to manage a connection instance
+ */
+
+#include "ofconnectionmanager_log.h"
+
+#include <sys/types.h>
+#include <sys/socket.h>
+#include <errno.h>
+#include <unistd.h>
+#include <string.h>
+
+#include "ofconnectionmanager_int.h"
+
+#include <SocketManager/socketmanager.h>
+#include <OFConnectionManager/ofconnectionmanager.h>
+
+#include <indigo/memory.h>
+#include <indigo/assert.h>
+
+#include <loci/loci_dump.h>
+
+#include <netinet/in.h>
+#include <netinet/tcp.h>
+#include <arpa/inet.h>
+#include <fcntl.h>
+
+#include "cxn_instance.h"
+
+
+/* Short hand logging macros */
+#define LOG_MSG(cxn, fmt, ...)                                  \
+    AIM_LOG_MSG("cxn %s: " fmt, cxn->desc, ##__VA_ARGS__)
+#define LOG_ERROR(cxn, fmt, ...)                                \
+    AIM_LOG_ERROR("cxn %s: " fmt, cxn->desc, ##__VA_ARGS__)
+#define LOG_WARN(cxn, fmt, ...)                                 \
+    AIM_LOG_WARN("cxn %s: " fmt, cxn->desc, ##__VA_ARGS__)
+#define LOG_INFO(cxn, fmt, ...)                                 \
+    AIM_LOG_INFO("cxn %s: " fmt, cxn->desc, ##__VA_ARGS__)
+#define LOG_VERBOSE(cxn, fmt, ...)                              \
+    AIM_LOG_VERBOSE("cxn %s: " fmt, cxn->desc, ##__VA_ARGS__)
+#define LOG_TRACE(cxn, fmt, ...)                                \
+    AIM_LOG_TRACE("cxn %s: " fmt, cxn->desc, ##__VA_ARGS__)
+#define LOG_INTERNAL(cxn, fmt, ...)                             \
+    AIM_LOG_INTERNAL("cxn %s(%p): " fmt, cxn->desc, cxn, ##__VA_ARGS__)
+
+
+/* Forward declaractions */
+static void
+cxn_state_set(connection_t *cxn, cxn_state_t new_state);
+static indigo_error_t
+cxn_try_to_connect(connection_t *cxn);
+
+
+/* Maximum number of messages to send per write callback */
+#define MAX_WRITE_MSGS 32
+
+
+/**
+ * Connection control blocks, indexed by connection index
+ */
+static connection_t connection[MAX_CONNECTIONS];
```

Comment:
For the record, I'd prefer these and `controller_t` to be dynamically allocated so that we can catch use-after-free bugs with valgrind.